app/services/bill-runs/two-part-tariff/prepare-licences-for-allocation.service.js

```diff
@@ -0,0 +1,142 @@
+ */
+function _abstractionOutsidePeriod (returnAbstractionPeriods, returnLine) {
+  const { startDate, endDate } = returnLine
+
+  return !periodsOverlap(returnAbstractionPeriods, [{ startDate, endDate }])
+}
+
+function _prepareChargeVersions (licence, billingPeriod) {
+  const { chargeVersions } = licence
+
+  chargeVersions.forEach((chargeVersion) => {
+    const { chargeReferences } = chargeVersion
+
+    _sortChargeReferencesBySubsistenceCharge(chargeReferences)
+    chargeVersion.chargePeriod = DetermineChargePeriodService.go(chargeVersion, billingPeriod)
+
+    chargeReferences.forEach((chargeReference) => {
+      const { chargeElements } = chargeReference
+
+      _prepareChargeElementsForMatching(chargeElements, chargeVersion.chargePeriod)
+    })
+  })
+}
+
+function _prepareChargeElementsForMatching (chargeElements, chargePeriod) {
+  chargeElements.forEach((chargeElement) => {
+    const {
+      abstractionPeriodStartDay,
+      abstractionPeriodStartMonth,
+      abstractionPeriodEndDay,
+      abstractionPeriodEndMonth
+    } = chargeElement
+
+    const abstractionPeriods = DetermineAbstractionPeriodService.go(
+      chargePeriod,
+      abstractionPeriodStartDay,
+      abstractionPeriodStartMonth,
+      abstractionPeriodEndDay,
+      abstractionPeriodEndMonth
+    )
+
+    chargeElement.returnLogs = []
+    chargeElement.allocatedQuantity = 0
+    chargeElement.abstractionPeriods = abstractionPeriods
+  })
+}
+
+async function _prepareReturnLogs (licence, billingPeriod) {
+  licence.returnLogs = await FetchReturnLogsForLicenceService.go(licence.licenceRef, billingPeriod)
+
+  _prepReturnsForMatching(licence.returnLogs, billingPeriod)
+}
+
+function _prepReturnsForMatching (returnLogs, billingPeriod) {
+  returnLogs.forEach((returnLog) => {
+    const { periodStartDay, periodStartMonth, periodEndDay, periodEndMonth, returnSubmissions } = returnLog
+    const abstractionPeriods = DetermineAbstractionPeriodService.go(
+      billingPeriod,
+      periodStartDay,
+      periodStartMonth,
+      periodEndDay,
+      periodEndMonth
+    )
+
+    let quantity = 0
+    let abstractionOutsidePeriod = false
+
+    returnSubmissions[0]?.returnSubmissionLines.forEach((returnSubmissionLine) => {
+      if (!abstractionOutsidePeriod) {
+        abstractionOutsidePeriod = _abstractionOutsidePeriod(abstractionPeriods, returnSubmissionLine)
+      }
+      returnSubmissionLine.unallocated = returnSubmissionLine.quantity / 1000
+      quantity += returnSubmissionLine.unallocated
+    })
+
+    returnLog.nilReturn = returnSubmissions[0]?.nilReturn ?? false
+    returnLog.quantity = quantity
+    returnLog.allocatedQuantity = 0
+    returnLog.abstractionPeriods = abstractionPeriods
+    returnLog.abstractionOutsidePeriod = abstractionOutsidePeriod
+    returnLog.matched = false
+  })
+}
+
+/**
+ * Sorts charge references by subsistence charge in descending order
+ * @param {Object[]} chargeReferences - Charge references to sort
```

Comment:
```suggestion
 * Sorts charge references by their subsistence charge in descending order
 * 
 * The rules for matching returns to charge elements require us to consider those charge elements assigned to the
 * charge reference with the highest subsistence charge first.
 * 
 * Normally, we'd deal with the ordering of things in the original fetch service. But to do that in this case would have
 * meant either creating a complex Knex-based query instead of just using Objection.js or, looping through the results
 * both in `FetchChargeVersionsService` and here.
```